File: solutions/Binary Search/4_median-of-two-sorted-arrays.py
from typing import List

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1

        m, n = len(nums1), len(nums2)
        pivot = (m + n) // 2
        l, r = 0, m - 1

        while True:
            i = l + (r-l) // 2
            j = pivot - i - 2

            left1 = nums1[i] if i >= 0 else float("-inf")
            left2 = nums2[j] if j >= 0 else float("-inf")
            right1 = nums1[i + 1] if (i + 1) < m else float("inf")
            right2 = nums2[j + 1] if (j + 1) < n else float("inf")

            if left1 <= right2 and left2 <= right1:
                if (m + n) % 2 == 0:
                    return (max(left1, left2) + min(right1, right2)) / 2

                else:
                    return min(right1, right2)

            elif right1 < left2:
                l = i + 1
            else:
                r = i - 1

# Time Complexity: O(log min(m, n))
# Space Complexity: O(1)


class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        res = []

        i, j = 0, 0
        while i < len(nums1) and j < len(nums2):
            if nums1[i] < nums2[j]:
                res.append(nums1[i])
                i += 1

            else:
                res.append(nums2[j])
                j += 1

        if i < len(nums1):
            res.extend(nums1[i:])
        if j < len(nums2):
            res.extend(nums2[j:])

        n = len(res)
        if n % 2 == 0:
            return (res[n//2 - 1] + res[n//2]) / 2
        else:
            return res[n//2]
        
# Time Complexity: O(m + n)
# Space Complexity: O(m + n)


from statistics import median

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:

        return median(sorted(nums1 + nums2))
    
# Time Complexity: O((m + n) log(m + n))
# Space Complexity: O(m + n)


import heapq

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        res = list(heapq.merge(nums1, nums2))
        n = len(res)
        if n % 2 == 0:
            return (res[n//2 - 1] + res[n//2]) / 2
        else:
            return res[n//2]
        
# Time Complexity: O(m + n)
# Space Complexity: O(m + n)


class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        res = sorted(nums1 + nums2)
        n = len(res)
        if n % 2 == 0:
            return (res[n//2 - 1] + res[n//2]) / 2
        else:
            return res[n//2]
        
# Time Complexity: O((m + n) log(m + n))
# Space Complexity: O(m + n)

File: solutions/Binary Search/4_median-of-two-sorted-arrays.cpp
#include <vector>
#include <climits>

using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        size_t m = nums1.size();
        size_t n = nums2.size();
        vector<int> res(m + n, 0);

        int i = 0, j = 0, k = 0;
        while (i < m && j < n) {
            if (nums1[i] < nums2[j]) res[k++] = nums1[i++];

            else res[k++] = nums2[j++];
        }
        while (i < m) res[k++] = nums1[i++];
        while (j < n) res[k++] = nums2[j++];

        if ((m + n) % 2) {
            return res[(m + n) / 2];
        } else {
            return (res[(m + n) / 2] + res[(m + n) / 2 - 1]) / 2.0;
        }
    }
};


class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        if (nums1.size() > nums2.size()) {
            swap(nums1, nums2);
        }

        int m = nums1.size();
        int n = nums2.size();
        int left = (m + n) / 2;
        bool isOdd = (m + n) % 2;

        int l = 0, r = min(m, left);
        while (l <= r) {
            int mid = (l + r + 1) / 2;

            int i = mid - 1;
            int j = left - mid - 1;

            int t1 = (i < 0)? INT_MIN : nums1[i];
            int t2 = (i + 1 < m)? nums1[i + 1] : INT_MAX;
            int b1 = (j < 0)? INT_MIN : nums2[j];
            int b2 = (j + 1 < n)? nums2[j + 1] : INT_MAX;

            if (t1 <= b2 && b1 <= t2) {
                if (isOdd) {
                    return min(t2, b2);
                } else {
                    return (max(t1, b1) + min(t2, b2)) / 2.0;
                }
            } else if (t1 > b2) r = mid - 1;
            else l = mid + 1;
        }

        return 0;
    }
};